### PR TITLE
Add query parameter to API endpoints

### PR DESCRIPTION
**Pull request by GGP1**
|Related issue|
|---|
|Closes https://github.com/wazuh/wazuh/issues/16714|

## Description

Added the query (`q`) parameter to several endpoints, updated their tests and specifications.

## Tests

```console
(venv) gasti@pop-os:~/work/wazuh$ python3 -m pytest api/api --disable-warnings
==================================================================================== test session starts ====================================================================================
platform linux -- Python 3.9.16, pytest-5.4.3, py-1.11.0, pluggy-0.13.1
rootdir: /home/gasti/work/wazuh/api
plugins: aiohttp-0.3.0, anyio-3.6.2, tavern-1.2.2, trio-0.7.0, html-2.1.1, asyncio-0.15.1, metadata-2.0.4
collected 544 items                                                                                                                                                                         

api/api/controllers/test/test_active_response_controller.py .                                                                                                                         [  0%]
api/api/controllers/test/test_agent_controller.py ...........................................                                                                                         [  8%]
api/api/controllers/test/test_cdb_list_controller.py ......                                                                                                                           [  9%]
api/api/controllers/test/test_ciscat_controller.py .                                                                                                                                  [  9%]
api/api/controllers/test/test_cluster_controller.py ........................                                                                                                          [ 13%]
api/api/controllers/test/test_decoder_controller.py .......                                                                                                                           [ 15%]
api/api/controllers/test/test_default_controller.py .                                                                                                                                 [ 15%]
api/api/controllers/test/test_experimental_controller.py ...............                                                                                                              [ 18%]
api/api/controllers/test/test_manager_controller.py ..................                                                                                                                [ 21%]
api/api/controllers/test/test_mitre_controller.py .......                                                                                                                             [ 22%]
api/api/controllers/test/test_overview_controller.py .                                                                                                                                [ 22%]
api/api/controllers/test/test_rootcheck_controller.py ....                                                                                                                            [ 23%]
api/api/controllers/test/test_rule_controller.py ........                                                                                                                             [ 25%]
api/api/controllers/test/test_sca_controller.py ..                                                                                                                                    [ 25%]
api/api/controllers/test/test_security_controller.py ...................................................                                                                              [ 34%]
api/api/controllers/test/test_syscheck_controller.py ....                                                                                                                             [ 35%]
api/api/controllers/test/test_syscollector_controller.py .........                                                                                                                    [ 37%]
api/api/controllers/test/test_task_controller.py .                                                                                                                                    [ 37%]
api/api/controllers/test/test_vulnerability_controller.py ....                                                                                                                        [ 38%]
api/api/models/test/test_model.py ...........................                                                                                                                         [ 43%]
api/api/test/test_alogging.py ..................                                                                                                                                      [ 46%]
api/api/test/test_authentication.py ...........                                                                                                                                       [ 48%]
api/api/test/test_configuration.py ............................................                                                                                                       [ 56%]
api/api/test/test_encoder.py ...                                                                                                                                                      [ 56%]
api/api/test/test_middlewares.py ............                                                                                                                                         [ 59%]
api/api/test/test_uri_parser.py ...                                                                                                                                                   [ 59%]
api/api/test/test_util.py ..............................................                                                                                                              [ 68%]
api/api/test/test_validator.py ...................................................................................................................................................... [ 95%]
.......................                                                                                                                                                               [100%]

============================================================================= 544 passed, 22 warnings in 2.10s ==============================================================================

```

### CODE REVIEW

**Review by EduLeon12 — COMMENTED** [review]
One concern for me regarding the issue is the naming of the parameter, even tho its well explained in the documentation of all the methods where its called, I think this should be reviewed to be explicitly called **query**, unless there is a reason behind the naming, since its being used or passed on to different methods. This could help to give context and leaves no room for confusion.

Regarding the issue, the scope is fulfilled, the solution is properly tested and well executed. Therefore it LGTM!

**Comment by Desvelao** [comment]
I see you did not add the request to the endpoint related to security rules. I missed the endpoint request of `GET /security/role_mapping` in https://github.com/wazuh/wazuh/issues/16552, it is GET `/security/rules` instead.

If you find some error in the request or there is a problem with some endpoint, we expect to get some notification of this.

**Review by Desvelao — CHANGES_REQUESTED** [review]
I think I missed the endpoint request. I reviewed it and this is `GET /security/rules` instead.

Please, if you find some error in the request, notify me before resolving the pull request or closing the issue as done.

**Review by RamosFe — DISMISSED** [review]
LGTM 🚀

File: framework/wazuh/rule.py
Comment (Selutario) [inline framework/wazuh/rule.py:211]: When using `q` with any field different than `name`, I'm getting an error. Could you check this out?

> GET /groups?q=mergedSum%3Ddefault

```JSON
{
  "title": "Bad Request",
  "detail": "Field does not exist.: Available fields: name. Field: mergedSum",
  "dapi_errors": {
    "master-node": {
      "error": "Field does not exist.: Available fields: name. Field: mergedSum"
    }
  },
  "error": 1408
}
```

These are common returned fields:
```JSON
{
  "data": {
    "affected_items": [
      {
        "name": "default",
        "count": 2,
        "mergedSum": "4a8724b20dee0124ff9656783c490c4e",
        "configSum": "ab73af41699f13fdd81903b5f23d8d00"
      }
    ],
    "total_affected_items": 1,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "All selected groups information was returned",
  "error": 0
}
```

If the fix is not that simple, open a new issue to improve the `process_array` function.

**Review by Selutario — CHANGES_REQUESTED** [review]

File: framework/wazuh/rule.py
Comment (GGP1) [reply framework/wazuh/rule.py:211]: This is expected. The `group` table has only one field (`name`) so that's the only one we can use to query rows:

https://github.com/wazuh/wazuh/blob/a7b092565dbd3f48e64bcfa77d8d2087431b0fb0/framework/wazuh/core/agent.py#L273

The `mergedSum` field is not in the table and is added after the query:

https://github.com/wazuh/wazuh/blob/a7b092565dbd3f48e64bcfa77d8d2087431b0fb0/framework/wazuh/agent.py#L591-L607

This is the  exact same problem I faced with the `group` and `group_hash` [issue](https://github.com/wazuh/wazuh/issues/16285). The filtering, sorting and selection is tied up to the fields that the table of the object has and other fields are rejected.

Comment (Selutario) [reply framework/wazuh/rule.py:211]: As we agreed after discussing, we can't add these filters to the endpoint if they won't work. I'm moving this to `On hold` until it's also discussed with the Frontend team and something is decided.

**Review by Selutario — DISMISSED** [review]
LGTM, moving it to blocked until merging https://github.com/wazuh/wazuh/pull/17703 and rebasing.

**Comment by Selutario** [comment]
Moving the issue back to on hold after @Desvelao request in https://github.com/wazuh/wazuh/issues/16552#issuecomment-1624859860. We may have to consider supporting `q` only for `name` field or using `process_array` for the database response. 

This should in any case be a temporary solution. The possibility of the `group` table including the `config_sum` and `merged_sum` fields should be discussed with the core team to avoid calculating it on demand, which is inefficient. This is already done in the `agent` table for agent groups and multigroups, since this issue:
- https://github.com/wazuh/wazuh/issues/12361

**Review by Selutario — APPROVED** [review]
LGTM! Moving to blocked until merging https://github.com/wazuh/wazuh/pull/17703 in all branches and rebasing. 

- **Important:** The PR should point to branch `4.6.0` instead of `master`.

**Comment by Selutario** [comment]
Done in https://github.com/wazuh/wazuh/pull/17865

**Pull request closed without merging**
